### PR TITLE
feat(adapter): Add patch data type to adapters and refactor AdapterBase usage

### PR DESCRIPTION
**Pull request by daffl**
This pull request adds the ability to use a `PatchData` type in the adapters which may be different than the usual data type.

It also refactors the `AdapterBase` to remove the confusing and convoluted `$` prefixed service methods. Instead an adapter just implements the underscored service methods as before. While for the top level usage backwards compatible this may be a breaking change in some hopefully still uncommon instances but it reduces a lot of the complexity when implementing a database adapter.

You can still extend directly from the adapter class to implement the actual service methods with your own signatures and return types. This will circumvent the legacy query sanitisation which is now done with query schemas anyway.

### CODE REVIEW

File: packages/adapter-commons/test/service.test.ts [hunk 1]
```diff
@@ -154,7 +154,7 @@ describe('@feathersjs/adapter-commons/service', () => {
     })
   })
 
-  it('allowsMulti', () => {
+  it.only('allowsMulti', () => {
```
Comment (marshallswain) [inline packages/adapter-commons/test/service.test.ts]: `it.only`

File: packages/knex/src/index.ts [hunk 2]
```diff
@@ -16,32 +17,55 @@ export class KnexService<T = any, D = Partial<T>, P extends Params<any> = KnexAd
...
+      ...params,
+      query: await this.sanitizeQuery(params)
+    })
   }
 
   async create(data: D, params?: P): Promise<T>
   async create(data: D[], params?: P): Promise<T[]>
   async create(data: D | D[], params?: P): Promise<T | T[]> {
+    if (Array.isArray(data) && !this.allowsMulti('create', params)) {
+      throw new MethodNotAllowed('Can not create multiple entries')
```
Comment (marshallswain) [inline packages/knex/src/index.ts:37]: "Can Not" --> "Cannot"

**Review by marshallswain — APPROVED** [review]
Looks good.  Will we make `sanitizeQuery` conditionally run in another PR, or did I miss it in this one?

**Comment by daffl** [comment]
I just added it in https://github.com/feathersjs/feathers/pull/2906/files#diff-686d31b4a5c93b9ab0ae6b60817ed194db3eccfaff344cfec6382f2439ce1e03R100

**Review by marshallswain — APPROVED** [review]
New updates look good.

**Pull request merged**